### PR TITLE
Added Topology Surface Charge SCI-1404

### PR DESCRIPTION
**Pull request by Alex-AMC**
- Added the ability to calculate the topology surface charge, which accounts for the node area % from the total surface. This better represents the available surface charge based on the atom position and steric hindrance.

### CODE REVIEW

**Review by pmbulit — APPROVED** [review]

File: scripts/surface_charge/surface_charge_calculator.py [hunk 1]
```diff
@@ -29,30 +28,107 @@ def __init__(self, crystal, use_existing_charges: bool = False):
         self.crystal = crystal
         self.surface = None
         self._surface_charge = None
+        self.surface_atom_charge = np.nan
+        self.node_projected_charge = np.nan
+        self.node_representative_charge = np.nan
```
Comment (psacchi-ccdc) [inline scripts/surface_charge/surface_charge_calculator.py:33]: what's the difference from initiating with "None"?

File: scripts/surface_charge/surface_charge_calculator.py [hunk 2]
```diff
@@ -29,30 +28,107 @@ def __init__(self, crystal, use_existing_charges: bool = False):
...
+        return np.round(np.sum([atom.partial_charge for atom in atoms]), 3)
+
+    def get_average_node_charge(self):
+        self.node_charge_dictionary = {}
+        node_list = list(self.surface.topology.nodes)
+        for node, atoms in self.surface.surface_node_atom_contacts.items():
+            node_index = node_list.index(node)
+            average_node_charge = 0
+            if len(atoms) > 0:
+                average_node_charge = self.sum_atom_charge(atoms)
```
Comment (psacchi-ccdc) [inline scripts/surface_charge/surface_charge_calculator.py:46]: the average is a sum?

File: scripts/surface_charge/surface_charge_calculator.py [hunk 3]
```diff
@@ -29,30 +28,107 @@ def __init__(self, crystal, use_existing_charges: bool = False):
...
+    def calculate_triangles_properties(self,
+                                       tri_index: List[Tuple[int, int, int]]) -> Dict[
+        Tuple[int, int, int], Dict[str, float]]:
+        surface_area = self.surface.descriptors.surface_area
+        self.triangles_properties = {}
+        triangle_areas = self.calculate_area_of_triangles(list(self.surface.topology.triangles))
+        total_triangle_area = sum(triangle_areas)
+        for node_index, triangle_area in zip(tri_index, triangle_areas):
+            average_triangle_charge = np.mean([self.node_charge_dictionary[i] for i in node_index])
+            triangle_representation = triangle_area / surface_area
```
[the remark this answers is not in the document]
Comment (psacchi-ccdc) [reply scripts/surface_charge/surface_charge_calculator.py:58]: actually I am still confused, probably would name the " average_node_charge = self.sum_atom_charge(atoms)" 
total_node_charge instead?

File: scripts/surface_charge/surface_charge_calculator.py [hunk 4]
```diff
@@ -29,30 +28,107 @@ def __init__(self, crystal, use_existing_charges: bool = False):
...
+        """ Calculates area of individual triangles from node positions using Heron's formula"""
+        triangle_areas = []
+        for triangle in triangles:
+            pos_0, pos_1, pos_2 = np.array(triangle[0]), np.array(triangle[1]), np.array(triangle[2]),
+            a_dist = self.calculate_length(pos_0, pos_1)
+            b_dist = self.calculate_length(pos_0, pos_2)
+            c_dist = self.calculate_length(pos_1, pos_2)
+            triangle_areas.append(self.compute_triangle_area(a_dist, b_dist, c_dist))
+
+        return triangle_areas
```
Comment (ckingsbury-ccdc) [inline scripts/surface_charge/surface_charge_calculator.py:101]: `
# the formula for determining the area of a triangle of three points in 3 dimensions
def area_tri(a, b, c):
    return np.linalg.norm(np.cross((b - a), (c - a))) / 2


# the area of a polygon using the above formula
def area_poly(arr):
    a = arr[1:-1] - arr[0]
    b = arr[2:] - arr[0]
    return np.sum(np.linalg.norm(np.cross(a, b, axis=1), axis=1)) / 2`

much more elegant

Comment (ckingsbury-ccdc) [reply scripts/surface_charge/surface_charge_calculator.py:101]: this is from the voronoi code, btw

**Review by ckingsbury-ccdc — APPROVED** [review]
see triangles

File: scripts/surface_charge/surface_charge_calculator.py (hunk 1 above)
Comment (Alex-AMC) [reply scripts/surface_charge/surface_charge_calculator.py:33]: As far as I am aware, numpy handles `np.nan` much better than None when you do an operation like `np.mean`

**Review by pmbulit — APPROVED** [review]

**Pull request merged**